core/src/main/java/org/apache/iceberg/rest/ScanTaskIterable.java

```diff
@@ -0,0 +1,249 @@
+        throw new RuntimeException("Worker failed processing planTask", e);
+      } finally {
+        int remaining = activeWorkers.decrementAndGet();
+
+        if (remaining == 0
+            && !planTasks.isEmpty()
+            && !shutdown.get()
+            && !initialFileScanTasks.isEmpty()) {
+          executorService.execute(new PlanTaskWorker());
+        }
+      }
+    }
+
+    private void processPlanTask(String planTask) throws InterruptedException {
+      FetchScanTasksResponse response = fetchScanTasks(planTask);
+      // immediately add any new plan tasks to the queue so the idle workers can pick them up
+      if (response.planTasks() != null) {
+        planTasks.addAll(response.planTasks());
+      }
+
+      // before blocking on the task queue, check for shutdown again
+      if (shutdown.get()) {
+        return;
+      }
+
+      // Now since the network IO is done, first add any initial file scan tasks
+      while (!initialFileScanTasks.isEmpty()) {
+        FileScanTask initialFileScanTask = initialFileScanTasks.poll();
+        if (initialFileScanTask != null) {
+          taskQueue.put(initialFileScanTask);
+        }
+      }
+
+      if (response.fileScanTasks() != null) {
+        for (FileScanTask task : response.fileScanTasks()) {
+          taskQueue.put(task);
+        }
+      }
+    }
+
+    private FetchScanTasksResponse fetchScanTasks(String planTask) {
+      FetchScanTasksRequest request = new FetchScanTasksRequest(planTask);
+
+      return client.post(
+          resourcePaths.fetchScanTasks(tableIdentifier),
+          request,
+          FetchScanTasksResponse.class,
+          headers,
+          ErrorHandlers.planTaskHandler(),
+          stringStringMap -> {},
+          parserContext);
+    }
+  }
+
+  private class ScanTasksIterator implements CloseableIterator<FileScanTask> {
+    private FileScanTask nextTask = null;
+
+    @Override
+    public boolean hasNext() {
+      if (nextTask != null) {
+        return true;
+      }
+
+      while (true) {
+        if (isDone()) {
+          return false;
+        }
+
+        try {
+          nextTask = taskQueue.poll(QUEUE_POLL_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+          if (nextTask != null) {
+            return true;
+          }
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          return false;
+        }
+      }
+    }
+
+    @Override
+    public FileScanTask next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException("No more scan tasks available");
+      }
+      FileScanTask result = nextTask;
+      nextTask = null;
+      return result;
+    }
+
+    @Override
+    public void close() {
+      shutdown.set(true);
+      LOG.info(
+          "ScanTasksIterator is closing. Clearing {} queued tasks and {} plan tasks.",
+          taskQueue.size(),
+          planTasks.size());
+      taskQueue.clear();
+      planTasks.clear();
+    }
+
+    private boolean isDone() {
```

Comment:
@singhpk234 is there a window where isDone is checked where initialFileScanTasks is polled but not yet offered to the taskQueue, and as a result we incorrectly return isDone is true? We'll probably need some more synchronization there